src/compiler/checker.ts

```diff
@@ -21268,18 +21318,52 @@ namespace ts {
             return true;
         }
 
+        function isReadonlyAssignmentDeclaration(d: Declaration) {
+            if (!isCallExpression(d)) {
+                return false;
+            }
+            if (!isBindableObjectDefinePropertyCall(d)) {
+                return false;
+            }
+            const objectLitType = checkExpressionCached(d.arguments[2]);
+            const valueType = getTypeOfPropertyOfType(objectLitType, "value" as __String);
+            if (valueType) {
+                const writableType = getTypeOfPropertyOfType(objectLitType, "writable" as __String);
+                if (!writableType || writableType === falseType || writableType === regularFalseType) {
+                    return true;
+                }
+                // We include this definition whereupon we walk back and check the type at the declaration because
+                // The usual definition of `Object.defineProperty` will _not_ cause literal type to be preserved in the
+                // argument types, should the type be contextualized by the call itself.
+                const writableProp = getPropertyOfType(objectLitType, "writable" as __String);

```

Comment:
would it make more sense to move this before `const writableType = getTypeOfPropertyOfType(...)` and then write `const writableType = getTypeOfSymbol(writableProp)`?